### ISSUE TITLE
Resuming on a different thread

### ISSUE THREAD

**Issue by zpyatt** [olk/P0876#34]
Oliver,

1st I appreciate your awesome work on both this proposal, boost.fiber, context, and coroutines. I've used and appreciated your work for years. 2nd please forgive my frustration and any strong language that follows. I am so disgusted/frustrated by the decision to remove resumption on a different thread. It seems to me this kills work sharing/stealing algorithms? It seems like a function referencing TLS would be an edge, ill-advised use-case. Why design for the exception and not for the rule? Why couldn't you have a more restricted std::immovable_fiber_context in addition to std::fiber_context, and just #@#% document the fact a function shouldn't reference TLS if it is going to be run on a std::fiber_context? This destroys the best use-cases of fibers for a 1% use-case. I'm so frustrated with the committee, its tempting to get away from C++. I hate it bad enough that they standardized 3 new keywords to support one !@#%& monad (the continuation monad), when we could have basically had DO notation for all monads like optional/error/continuation/etc.... Plus your stackful coroutines proposals were way better. End rant. Anyway is there anything to be done about this BS?

**Comment by olk** [comment]
Unfortunately compilers are permitted to use TLS for optimization (key is stored at the stack), e.g. only preventing explicit usage of TLS in the function that runs on a fiber and might be migrated to another thread is still dangerous.
I've had the idea to change the way TLS works - in fact the implementations I saw on UNIX are lookup lists. It might be possible to modify it in such a way that the TLS key relates to the correct value.
Unfortunately I don't get much support from the C++ community/committee and I've not enough time (have to do it in my free time).

**Comment by zpyatt** [comment]
Wow. 1st I guess I should know what I'm talking about before getting so irritated. So are you saying that even if my function doesn't "explicitly" make use of TLS, the compile might choose to store some of my function's data in TLS? Wow. That's a bummer, as I don't see a way of "fixing" fibers if that's the case. Actually, that breaks fibers, which is a shame. How does boost.fiber or context work if this is the case? I think this breaks fibers so bad it isn't even worth your time for the proposal. What advantage would std::fiber have over Boost? Anyway, thanks or you time and hard work, its brilliant, brings C++ an infinitesimally tiny little bit closer to the expressiveness of Lisp.

**Comment by zpyatt** [comment]
oh feel free to close this.

**Comment by olk** [comment]
Yes, the compiler might choose TLS for optimization - it isn't a problem for fibers until migrating them between threads.
But it can be solved - TLS implementations use a key which is in fact an address in a global lookup table. This TLS key is used to access the thread-specific data. If this key would be reloaded/recalculated before each access - no failures with fibers would be caused. But this would be slow down the performance. For optimization the TLS-key is stored at the stack.
As you know each fiber has its own stack (you can track thread stacks as fiber stacks == each thread has at least one fiber running). If the key would be fiber-related (computation contains fiber == stack specific data) we would not encounter problems if the fiber is migrated between threads.

**Comment by zpyatt** [comment]
I think you mentioned this in a "discussion" with Gor on google groups a few years ago, I assume the "discussion" with the committee went about as well? Why didn't they go for it? 

Good luck, is there any way to hep with this? I wish features had a public debate/forum that the committee could at least take under advisement, not that we would democratize standardization, but at least public input/feedback. It seems like its been one (many) ridiculous decisions after another lately.

**Comment by olk** [comment]
+1
